core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1338,16 +1340,21 @@ protected SqlNode performUnconditionalRewrites(
         assert call instanceof SqlBasicCall;
         final SqlUnresolvedFunction function =
             (SqlUnresolvedFunction) call.getOperator();
-        // This function hasn't been resolved yet.  Perform
-        // a half-hearted resolution now in case it's a
-        // builtin function requiring special casing.  If it's
-        // not, we'll handle it later during overload resolution.
-        final List<SqlOperator> overloads = new ArrayList<>();
-        opTab.lookupOperatorOverloads(function.getNameAsId(),
-            function.getFunctionType(), SqlSyntax.FUNCTION, overloads,
-            catalogReader.nameMatcher());
-        if (overloads.size() == 1) {
-          ((SqlBasicCall) call).setOperator(overloads.get(0));
+        SqlFunctionCategory category = function.getFunctionType();
+        // Don't want to overwrite macros since they lose their identifiable
+        // category.
+        if (category == null || !category.isUserDefinedMacro()) {
```

Comment:
This is necessary because `lookupOperatorOverloads` will convert this a SqlUserDefinedFunction with category USER_DEFINED_FUNCTION. Ideally the category of this one is set to USER_DEFINED_MACRO on creation, but the constructor that allows that is protected: https://github.com/googleinterns/calcite/blob/97399801253a1417ff2676c8740c06180b2e26f1/core/src/main/java/org/apache/calcite/sql/validate/SqlUserDefinedFunction.java#L56.

Is it alright to make this public?

Reply:
SGTM